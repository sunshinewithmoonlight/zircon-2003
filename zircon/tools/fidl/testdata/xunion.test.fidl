library fidl.test.json;

xunion FlexibleFoo {
    1: string s;
    2: int32 i;
};

strict xunion StrictFoo {
    1: string s;
    2: int32 i;
};

xunion ExplicitFoo {
    2: string s;
    1: int32 i;
    3: reserved;
};

strict xunion ExplicitStrictFoo {
    1: reserved;
    3: string s;
    2: int32 i;
};

xunion OlderSimpleUnion {
    1: int64 i;
    2: float32 f;
};

xunion NewerSimpleUnion {
    // float f;  // removed

    1: int64 i; // unchanged

    2: string s; // added
    3: vector<string> v; // added
};

// 1. This xunion is currently unused by code, and for tracking changes in the goldens only.
// 2. There's no FlexibleSimpleXUnion, since that's covered by OlderSimpleUnion above.
strict xunion StrictSimpleXUnion {
    1: int32 i;
    2: float32 f;
    3: string s;
};

struct Empty {
};

xunion XUnionContainingEmptyStruct {
    // This is a test to prevent regressions for FIDL-762. Note that we need FIDL-758 to be resolved
    // to ensure this actually works, since goldens aren't currently compiled.
    1: Empty empty;
};

strict xunion StrictBoundedXUnion {
    1: vector<uint8>:10 v;
};

protocol TestProtocol {
    StrictXUnionHenceResponseMayBeStackAllocated() -> (StrictBoundedXUnion xu);
    FlexibleXUnionHenceResponseMustBeHeapAllocated() -> (OlderSimpleUnion xu);
};

struct StructWithNullableXUnion {
    OlderSimpleUnion? x1;
};

xunion ExplicitXUnion {
    3: reserved;
    2: reserved;
    1: int64 i;
    4: float32 f;
};
