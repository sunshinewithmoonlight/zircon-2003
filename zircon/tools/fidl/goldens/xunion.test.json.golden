{
  "version": "0.0.1",
  "name": "fidl.test.json",
  "library_dependencies": [],
  "bits_declarations": [],
  "const_declarations": [],
  "enum_declarations": [],
  "interface_declarations": [
    {
      "name": "fidl.test.json/TestProtocol",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 60,
        "column": 10
      },
      "methods": [
        {
          "ordinal": 7548006651618394112,
          "generated_ordinal": 8255078252057434586,
          "name": "StrictXUnionHenceResponseMayBeStackAllocated",
          "location": {
            "filename": "xunion.test.fidl",
            "line": 61,
            "column": 5
          },
          "has_request": true,
          "maybe_request": [],
          "maybe_request_type_shape_v1": {
            "inline_size": 16,
            "alignment": 8,
            "depth": 0,
            "max_handles": 0,
            "max_out_of_line": 0,
            "has_padding": false,
            "has_flexible_envelope": false,
            "contains_union": false
          },
          "has_response": true,
          "maybe_response": [
            {
              "type": {
                "kind": "identifier",
                "identifier": "fidl.test.json/StrictBoundedXUnion",
                "nullable": false
              },
              "name": "xu",
              "location": {
                "filename": "xunion.test.fidl",
                "line": 61,
                "column": 76
              },
              "field_shape_v1": {
                "offset": 16,
                "padding": 0
              }
            }
          ],
          "maybe_response_payload": "fidl.test.json/SomeLongAnonymousPrefix1",
          "maybe_response_type_shape_v1": {
            "inline_size": 40,
            "alignment": 8,
            "depth": 2,
            "max_handles": 0,
            "max_out_of_line": 32,
            "has_padding": true,
            "has_flexible_envelope": false,
            "contains_union": true
          },
          "is_composed": false
        },
        {
          "ordinal": 1688296084360658944,
          "generated_ordinal": 1140178169303334883,
          "name": "FlexibleXUnionHenceResponseMustBeHeapAllocated",
          "location": {
            "filename": "xunion.test.fidl",
            "line": 62,
            "column": 5
          },
          "has_request": true,
          "maybe_request": [],
          "maybe_request_type_shape_v1": {
            "inline_size": 16,
            "alignment": 8,
            "depth": 0,
            "max_handles": 0,
            "max_out_of_line": 0,
            "has_padding": false,
            "has_flexible_envelope": false,
            "contains_union": false
          },
          "has_response": true,
          "maybe_response": [
            {
              "type": {
                "kind": "identifier",
                "identifier": "fidl.test.json/OlderSimpleUnion",
                "nullable": false
              },
              "name": "xu",
              "location": {
                "filename": "xunion.test.fidl",
                "line": 62,
                "column": 75
              },
              "field_shape_v1": {
                "offset": 16,
                "padding": 0
              }
            }
          ],
          "maybe_response_payload": "fidl.test.json/SomeLongAnonymousPrefix3",
          "maybe_response_type_shape_v1": {
            "inline_size": 40,
            "alignment": 8,
            "depth": 1,
            "max_handles": 0,
            "max_out_of_line": 8,
            "has_padding": true,
            "has_flexible_envelope": true,
            "contains_union": true
          },
          "is_composed": false
        }
      ]
    }
  ],
  "service_declarations": [],
  "struct_declarations": [
    {
      "name": "fidl.test.json/SomeLongAnonymousPrefix1",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 61,
        "column": 55
      },
      "anonymous": true,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.json/StrictBoundedXUnion",
            "nullable": false
          },
          "name": "xu",
          "location": {
            "filename": "xunion.test.fidl",
            "line": 61,
            "column": 76
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 32,
        "has_padding": true,
        "has_flexible_envelope": false,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/SomeLongAnonymousPrefix3",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 62,
        "column": 57
      },
      "anonymous": true,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.json/OlderSimpleUnion",
            "nullable": false
          },
          "name": "xu",
          "location": {
            "filename": "xunion.test.fidl",
            "line": 62,
            "column": 75
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/Empty",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 47,
        "column": 8
      },
      "anonymous": false,
      "members": [],
      "type_shape_v1": {
        "inline_size": 1,
        "alignment": 1,
        "depth": 0,
        "max_handles": 0,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "contains_union": false
      }
    },
    {
      "name": "fidl.test.json/StructWithNullableXUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 65,
        "column": 8
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.json/OlderSimpleUnion",
            "nullable": true
          },
          "name": "x1",
          "location": {
            "filename": "xunion.test.fidl",
            "line": 66,
            "column": 23
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    }
  ],
  "table_declarations": [],
  "union_declarations": [
    {
      "name": "fidl.test.json/FlexibleFoo",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 3,
        "column": 8
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 4,
            "column": 15
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 5,
            "column": 14
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/StrictFoo",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 8,
        "column": 15
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 9,
            "column": 15
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 10,
            "column": 14
          }
        }
      ],
      "strict": true,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/ExplicitFoo",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 13,
        "column": 8
      },
      "members": [
        {
          "ordinal": 2,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 14,
            "column": 15
          }
        },
        {
          "ordinal": 1,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 15,
            "column": 14
          }
        },
        {
          "ordinal": 3,
          "reserved": true,
          "location": {
            "filename": "xunion.test.fidl",
            "line": 16,
            "column": 5
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/ExplicitStrictFoo",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 19,
        "column": 15
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": true,
          "location": {
            "filename": "xunion.test.fidl",
            "line": 20,
            "column": 5
          }
        },
        {
          "ordinal": 3,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 21,
            "column": 15
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 22,
            "column": 14
          }
        }
      ],
      "strict": true,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/OlderSimpleUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 25,
        "column": 8
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int64"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 26,
            "column": 14
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "f",
          "type": {
            "kind": "primitive",
            "subtype": "float32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 27,
            "column": 16
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/NewerSimpleUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 30,
        "column": 8
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int64"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 33,
            "column": 14
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 35,
            "column": 15
          }
        },
        {
          "ordinal": 3,
          "reserved": false,
          "name": "v",
          "type": {
            "kind": "vector",
            "element_type": {
              "kind": "string",
              "nullable": false
            },
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 36,
            "column": 23
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 3,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/StrictSimpleXUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 41,
        "column": 15
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 42,
            "column": 14
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "f",
          "type": {
            "kind": "primitive",
            "subtype": "float32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 43,
            "column": 16
          }
        },
        {
          "ordinal": 3,
          "reserved": false,
          "name": "s",
          "type": {
            "kind": "string",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 44,
            "column": 15
          }
        }
      ],
      "strict": true,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/XUnionContainingEmptyStruct",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 50,
        "column": 8
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "empty",
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.json/Empty",
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 53,
            "column": 14
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/StrictBoundedXUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 56,
        "column": 15
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "v",
          "type": {
            "kind": "vector",
            "element_type": {
              "kind": "primitive",
              "subtype": "uint8"
            },
            "maybe_element_count": 10,
            "nullable": false
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 57,
            "column": 25
          }
        }
      ],
      "strict": true,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 2,
        "max_handles": 0,
        "max_out_of_line": 32,
        "has_padding": true,
        "has_flexible_envelope": false,
        "contains_union": true
      }
    },
    {
      "name": "fidl.test.json/ExplicitXUnion",
      "location": {
        "filename": "xunion.test.fidl",
        "line": 69,
        "column": 8
      },
      "members": [
        {
          "ordinal": 3,
          "reserved": true,
          "location": {
            "filename": "xunion.test.fidl",
            "line": 70,
            "column": 5
          }
        },
        {
          "ordinal": 2,
          "reserved": true,
          "location": {
            "filename": "xunion.test.fidl",
            "line": 71,
            "column": 5
          }
        },
        {
          "ordinal": 1,
          "reserved": false,
          "name": "i",
          "type": {
            "kind": "primitive",
            "subtype": "int64"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 72,
            "column": 14
          }
        },
        {
          "ordinal": 4,
          "reserved": false,
          "name": "f",
          "type": {
            "kind": "primitive",
            "subtype": "float32"
          },
          "location": {
            "filename": "xunion.test.fidl",
            "line": 73,
            "column": 16
          }
        }
      ],
      "strict": false,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": true,
        "contains_union": true
      }
    }
  ],
  "type_alias_declarations": [],
  "declaration_order": [
    "fidl.test.json/StructWithNullableXUnion",
    "fidl.test.json/StrictSimpleXUnion",
    "fidl.test.json/StrictFoo",
    "fidl.test.json/StrictBoundedXUnion",
    "fidl.test.json/OlderSimpleUnion",
    "fidl.test.json/TestProtocol",
    "fidl.test.json/NewerSimpleUnion",
    "fidl.test.json/FlexibleFoo",
    "fidl.test.json/ExplicitXUnion",
    "fidl.test.json/ExplicitStrictFoo",
    "fidl.test.json/ExplicitFoo",
    "fidl.test.json/Empty",
    "fidl.test.json/XUnionContainingEmptyStruct"
  ],
  "declarations": {
    "fidl.test.json/TestProtocol": "interface",
    "fidl.test.json/SomeLongAnonymousPrefix0": "struct",
    "fidl.test.json/SomeLongAnonymousPrefix1": "struct",
    "fidl.test.json/SomeLongAnonymousPrefix2": "struct",
    "fidl.test.json/SomeLongAnonymousPrefix3": "struct",
    "fidl.test.json/Empty": "struct",
    "fidl.test.json/StructWithNullableXUnion": "struct",
    "fidl.test.json/FlexibleFoo": "union",
    "fidl.test.json/StrictFoo": "union",
    "fidl.test.json/ExplicitFoo": "union",
    "fidl.test.json/ExplicitStrictFoo": "union",
    "fidl.test.json/OlderSimpleUnion": "union",
    "fidl.test.json/NewerSimpleUnion": "union",
    "fidl.test.json/StrictSimpleXUnion": "union",
    "fidl.test.json/XUnionContainingEmptyStruct": "union",
    "fidl.test.json/StrictBoundedXUnion": "union",
    "fidl.test.json/ExplicitXUnion": "union"
  }
}
